components/ai_chat/core/browser/conversation_handler.cc

```diff
@@ -1255,20 +1338,53 @@ void ConversationHandler::OnHistoryUpdate() {
   for (auto& client : conversation_ui_handlers_) {
     client->OnConversationHistoryUpdate();
   }
-  OnConversationEntriesChanged();
 }
 
-void ConversationHandler::OnConversationEntriesChanged() {
+void ConversationHandler::OnConversationEntryRemoved(
+    std::optional<std::string> entry_uuid) {
+  OnHistoryUpdate();
+  if (!entry_uuid.has_value()) {
+    return;
+  }
   for (auto& observer : observers_) {
-    // TODO(petemill): only tell observers about complete turns. This is
-    // expensive to do for every event generated by in-progress turns,
-    // and consumers likely only need complete ones (e.g. database save).
-    std::vector<mojom::ConversationTurnPtr> history;
-    for (const auto& turn : chat_history_) {
-      history.emplace_back(turn->Clone());
+    observer.OnConversationEntryRemoved(this, entry_uuid.value());
+  }
+}
+
+void ConversationHandler::OnConversationEntryAdded(
+    mojom::ConversationTurnPtr& entry) {
+  // Only notify about staged entries once we have the first staged entry
+  if (entry->from_brave_search_SERP) {
+    OnHistoryUpdate();
+    return;
+  }
+  std::optional<std::string> associated_content_value;
+  if (is_content_different_ && associated_content_delegate_) {
+    associated_content_value =
+        associated_content_delegate_->GetCachedTextContent();
+  }
+  // If this is the first entry that isn't staged, notify about all previous
+  // staged entries
+  if (!entry->from_brave_search_SERP &&
+      base::ranges::all_of(chat_history_,
+                           [&entry](mojom::ConversationTurnPtr& history_entry) {
+                             return history_entry == entry ||
+                                    history_entry->from_brave_search_SERP;
+                           })) {
+    // Notify every item in chat history
+    for (auto& observer : observers_) {
+      for (auto& history_entry : chat_history_) {
+        observer.OnConversationEntryAdded(this, history_entry,
+                                          associated_content_value);
+      }
     }
-    observer.OnConversationEntriesChanged(this, std::move(history));
+    OnHistoryUpdate();
+    return;
   }
+  for (auto& observer : observers_) {
+    observer.OnConversationEntryAdded(this, entry, associated_content_value);
```

Comment:
So let's say we have two tabs both loaded with same url and we have two separate conversations associated to the contents accordingly. When we save entry, we store two different entries with the same content into db currently right?

Reply:
That's right. I don't think we can guarantee the same content at the same URL

Reply:
It's actually coming from https://github.com/brave/brave-core/pull/25876#discussion_r1809678781. I'm more inclined toward not saving the web contents at all because it can be easily outdated and every time we reload previous conversation, the web page has already been loaded or being loaded and we should just use that. 
The effort we spend on serialize/deserialize and encrypt/decrypt plus increasing disk usage outweigh the time we want to save doing extracting ex. fetching video transcript, waiting for long contents to be loaded or OCRing pdf/google doc IMHO.

Reply:
The web page won't have been loaded. Once the page is navigated away from or closed, we don't have the content anymore. We don't re-open the Tab and attempt to recreate the state the web content was in some time in the future when the user re-opens the page.

Reply:
oh I see, in that case I think we can have something like another mapping with key equals hash(content) and value equals content then we persist it to database

Reply:
Yes, avoid duplicating the storage.
Probably not worth it because I had the wrong calculation of approximate size it needs.